### DISCUSSION TITLE
How to correctly use GroupInfo for multithreading a pipeline task?

### DISCUSSION THREAD

**Discussion by Risiamu** [PowerShell/PowerShell#17757] [Q&A]
I was trying to use multithreading to boost a batch ImageMagick work.
I have about 50k photos and i want them to be the same size.
But i can't get it to work, the parameter passing for script-block is broken...
Here is the code:

```powershell
$pic = ls
$counter = [pscustomobject] @{ Value = 0 }
$groupSize = 4520
$groups = $pic | Group-Object -Property { [math]::Floor($counter.Value++ / $groupSize) }

foreach($group in $groups){
    $temp = $group.Group | sort-object;
    echo $temp.Gettype();
    echo $temp[0];
    Start-Job -ScriptBlock {Param([System.Object[]]$objects)
    
    $objects | magick $_.Name -resize 1920x1080 -background white -gravity center -extent 1920x1080 $_.Name} -ArgumentList (,$temp)
}
```
I been working on this for at least an hour, but i just don't know what is wrong.

**Comment by SeeminglyScience** [comment]
Probably:

```powershell
foreach ($object in $objects) {
	magick $object.Name -resize 1920x1080 -background white -gravity center -extent 1920x1080 $object.Name
}
```

The variable `$_` is not actually set at the time parameter binding happens for the native command. That's why when you see it used in the middle of a pipeline, it's in a `ScriptBlock` so execution is deferred until right before each pipeline item is processed. For a native command, there's no way around needing either `foreach` or `ForEach-Object`.

**Comment by kasini3000** [comment]
1 Processing image files requires cpu or gpu
2 ImageMagick,This software supports running on win or linux
3 This is a typical distribution file for distributed computing needs.
Mainly 2 functions are required:
* Copy some files from master to several nodes,
* Batch execution, batch processing.
* copy back.
Suggested solutions are welcome
